### ISSUE TITLE
bindlink value clash

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1176]
I'm getting the same hash value for 2 different non alpha-equivalent bindlinks. I know collision is impossible to avoid but since I'm getting that now I suspect the hash value algo isn't terribly good (either that or I'm incredibly unlucky).

The 2 clashing bindlinks are

```scheme
(BindLink
  (VariableList
    (TypedVariableLink
      (VariableNode "$B-a083600") ; [1075803195405159200][3]
      (TypeChoice
        (TypeNode "OrLink") ; [1641200931040822982][1]
        (TypeNode "AndLink") ; [5150993344963829271][1]
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "NotLink") ; [7029353635928836065][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [12553683925648135849][1]
    ) ; [12502753568736379187][3]
    (TypedVariableLink
      (VariableNode "$B-5560bed3") ; [954956885300767879][3]
      (TypeChoice
        (TypeNode "OrLink") ; [1641200931040822982][1]
        (TypeNode "AndLink") ; [5150993344963829271][1]
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "NotLink") ; [7029353635928836065][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [12553683925648135849][1]
    ) ; [17738197372146241402][3]
    (TypedVariableLink
      (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
      (TypeChoice
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [15286865261196305675][1]
    ) ; [16170795669022151701][3]
  ) ; [13184149590889680758][3]
  (AndLink
    (VariableNode "$B-a083600") ; [1075803195405159200][3]
    (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
    (NotLink
      (EqualLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
      ) ; [9530461658253061016][3]
    ) ; [9550780633133926061][3]
    (NotLink
      (IdenticalLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [9771779474766691712][3]
    ) ; [9792098449647556757][3]
    (EvaluationLink
      (GroundedPredicateNode "scm: gt-zero-confidence") ; [2096565403849494700][1]
      (EquivalenceLink
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
      ) ; [16346461327915047267][3]
    ) ; [13451195623684469822][3]
    (ImplicationLink
      (AndLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
      ) ; [17393623462368785383][3]
      (VariableNode "$B-5560bed3") ; [954956885300767879][3]
    ) ; [13870626034087997381][3]
    (ImplicationLink
      (VariableNode "$B-5560bed3") ; [954956885300767879][3]
      (LambdaLink
        (TypedVariableLink
          (VariableNode "$X") ; [6809909406030619949][1]
          (TypeNode "ConceptNode") ; [3788388250666256890][1]
        ) ; [17718894144153964593][1]
        (EvaluationLink
          (PredicateNode "take") ; [3396821867884253757][1]
          (ListLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (ConceptNode "compound-A") ; [5571266438078870282][1]
          ) ; [9339663655215837098][1]
        ) ; [12459373116713205302][1]
      ) ; [18413871478668594639][1]
    ) ; [14585749684933467565][3]
    (EvaluationLink
      (GroundedPredicateNode "scm: implication-construction-precondition") ; [7996446579149881740][1]
      (ListLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
      ) ; [17163728775145283731][3]
    ) ; [15273729081877186638][3]
    (NotLink
      (IdenticalLink
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [15601785458906142039][3]
    ) ; [15622104433787007084][3]
    (NotLink
      (IdenticalLink
        (VariableNode "$B-5560bed3") ; [954956885300767879][3]
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
      ) ; [16115899068622670104][3]
    ) ; [16136218043503535149][3]
    (EquivalenceLink
      (PredicateNode "take-treatment-1") ; [693477798751519767][1]
      (VariableNode "$B-a083600") ; [1075803195405159200][3]
    ) ; [16346461327915047267][3]
  ) ; [14137636326773825635][3]
  (ExecutionOutputLink
    (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
    (ListLink
      (ImplicationLink (stv 0,200000 0,576000)
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [15180311865663055677][1]
      (ExecutionOutputLink
        (GroundedSchemaNode "scm: equivalence-to-implication-formula") ; [657362063185547793][1]
        (ListLink
          (ImplicationLink
            (PredicateNode "take-treatment-1") ; [693477798751519767][1]
            (VariableNode "$B-a083600") ; [1075803195405159200][3]
          ) ; [16288987656109171854][3]
          (EquivalenceLink
            (PredicateNode "take-treatment-1") ; [693477798751519767][1]
            (VariableNode "$B-a083600") ; [1075803195405159200][3]
          ) ; [16346461327915047267][3]
        ) ; [10106419507727073668][3]
      ) ; [15383359922948197147][3]
      (ExecutionOutputLink
        (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
        (ListLink
          (ImplicationLink
            (VariableNode "$B-a083600") ; [1075803195405159200][3]
            (LambdaLink
              (TypedVariableLink
                (VariableNode "$X") ; [6809909406030619949][1]
                (TypeNode "ConceptNode") ; [3788388250666256890][1]
              ) ; [17718894144153964593][1]
              (EvaluationLink
                (PredicateNode "take") ; [3396821867884253757][1]
                (ListLink
                  (VariableNode "$X") ; [6809909406030619949][1]
                  (ConceptNode "compound-A") ; [5571266438078870282][1]
                ) ; [9339663655215837098][1]
              ) ; [12459373116713205302][1]
            ) ; [18413871478668594639][1]
          ) ; [9350305881523605350][3]
          (ExecutionOutputLink
            (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
            (ListLink
              (ImplicationLink
                (VariableNode "$B-a083600") ; [1075803195405159200][3]
                (VariableNode "$B-5560bed3") ; [954956885300767879][3]
              ) ; [10338135361865330206][3]
              (ExecutionOutputLink
                (GroundedSchemaNode "scm: implication-implicant-distribution-formula") ; [4181613870801572780][1]
                (ListLink
                  (ImplicationLink
                    (VariableNode "$B-a083600") ; [1075803195405159200][3]
                    (AndLink
                      (VariableNode "$B-a083600") ; [1075803195405159200][3]
                      (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
                    ) ; [17393623462368785383][3]
                  ) ; [17553429902078571902][3]
                  (ExecutionOutputLink
                    (GroundedSchemaNode "scm: implication-construction-formula") ; [6458395065017447120][1]
                    (ListLink
                      (ImplicationLink
                        (VariableNode "$B-a083600") ; [1075803195405159200][3]
                        (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
                      ) ; [18313202211262791991][3]
                      (VariableNode "$B-a083600") ; [1075803195405159200][3]
                      (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
                    ) ; [13737951571107005930][3]
                  ) ; [16758168272830515232][3]
                ) ; [15351232422213639985][3]
              ) ; [17024646009651502531][3]
              (ImplicationLink
                (AndLink
                  (VariableNode "$B-a083600") ; [1075803195405159200][3]
                  (VariableNode "$Q-5fa8b92b") ; [8930023734698229664][3]
                ) ; [17393623462368785383][3]
                (VariableNode "$B-5560bed3") ; [954956885300767879][3]
              ) ; [13870626034087997381][3]
            ) ; [13603448326412033977][3]
          ) ; [13068574681062347282][3]
          (ImplicationLink
            (VariableNode "$B-5560bed3") ; [954956885300767879][3]
            (LambdaLink
              (TypedVariableLink
                (VariableNode "$X") ; [6809909406030619949][1]
                (TypeNode "ConceptNode") ; [3788388250666256890][1]
              ) ; [17718894144153964593][1]
              (EvaluationLink
                (PredicateNode "take") ; [3396821867884253757][1]
                (ListLink
                  (VariableNode "$X") ; [6809909406030619949][1]
                  (ConceptNode "compound-A") ; [5571266438078870282][1]
                ) ; [9339663655215837098][1]
              ) ; [12459373116713205302][1]
            ) ; [18413871478668594639][1]
          ) ; [14585749684933467565][3]
        ) ; [16283650869652643608][3]
      ) ; [15748777224302956913][3]
    ) ; [14020798431337846428][3]
  ) ; [13485924785988159733][3]
) ; [15511925606899762288][3]
```

and

```scheme
(BindLink
  (VariableList
    (TypedVariableLink
      (VariableNode "$B-a083600") ; [1075803195405159200][3]
      (TypeChoice
        (TypeNode "OrLink") ; [1641200931040822982][1]
        (TypeNode "AndLink") ; [5150993344963829271][1]
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "NotLink") ; [7029353635928836065][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [12553683925648135849][1]
    ) ; [12502753568736379187][3]
    (TypedVariableLink
      (VariableNode "$B-79625a3") ; [7102392297705646207][3]
      (TypeChoice
        (TypeNode "OrLink") ; [1641200931040822982][1]
        (TypeNode "AndLink") ; [5150993344963829271][1]
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "NotLink") ; [7029353635928836065][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [12553683925648135849][1]
    ) ; [17689381170702158450][3]
    (TypedVariableLink
      (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
      (TypeChoice
        (TypeNode "LambdaLink") ; [6352458681478507967][1]
        (TypeNode "PredicateNode") ; [7305492803481351454][1]
      ) ; [15286865261196305675][1]
    ) ; [10785106554706430372][3]
  ) ; [15410897865773997821][3]
  (AndLink
    (VariableNode "$B-a083600") ; [1075803195405159200][3]
    (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
    (NotLink
      (EqualLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
      ) ; [9646755080148183879][3]
    ) ; [9667074055029048924][3]
    (NotLink
      (IdenticalLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [9771779474766691712][3]
    ) ; [9792098449647556757][3]
    (ImplicationLink
      (PredicateNode "take-treatment-1") ; [693477798751519767][1]
      (VariableNode "$B-79625a3") ; [7102392297705646207][3]
    ) ; [13092204721554883053][3]
    (EvaluationLink
      (GroundedPredicateNode "scm: gt-zero-confidence") ; [2096565403849494700][1]
      (EquivalenceLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$B-79625a3") ; [7102392297705646207][3]
      ) ; [16543044446076083947][3]
    ) ; [13647778741845506502][3]
    (EvaluationLink
      (GroundedPredicateNode "scm: implication-construction-precondition") ; [7996446579149881740][1]
      (ListLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
      ) ; [17280022197040406594][3]
    ) ; [15390022503772309501][3]
    (NotLink
      (IdenticalLink
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [15601785458906142039][3]
    ) ; [15622104433787007084][3]
    (EquivalenceLink
      (VariableNode "$B-a083600") ; [1075803195405159200][3]
      (VariableNode "$B-79625a3") ; [7102392297705646207][3]
    ) ; [16543044446076083947][3]
    (ImplicationLink
      (AndLink
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
        (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
      ) ; [17509916884263908246][3]
      (LambdaLink
        (TypedVariableLink
          (VariableNode "$X") ; [6809909406030619949][1]
          (TypeNode "ConceptNode") ; [3788388250666256890][1]
        ) ; [17718894144153964593][1]
        (EvaluationLink
          (PredicateNode "take") ; [3396821867884253757][1]
          (ListLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (ConceptNode "compound-A") ; [5571266438078870282][1]
          ) ; [9339663655215837098][1]
        ) ; [12459373116713205302][1]
      ) ; [18413871478668594639][1]
    ) ; [16720479476285327004][3]
    (NotLink
      (IdenticalLink
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (VariableNode "$B-a083600") ; [1075803195405159200][3]
      ) ; [16710461249352258216][3]
    ) ; [16730780224233123261][3]
  ) ; [9431942479485073351][3]
  (ExecutionOutputLink
    (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
    (ListLink
      (ImplicationLink (stv 0,200000 0,641520)
        (PredicateNode "take-treatment-1") ; [693477798751519767][1]
        (LambdaLink
          (TypedVariableLink
            (VariableNode "$X") ; [6809909406030619949][1]
            (TypeNode "ConceptNode") ; [3788388250666256890][1]
          ) ; [17718894144153964593][1]
          (EvaluationLink
            (PredicateNode "take") ; [3396821867884253757][1]
            (ListLink
              (VariableNode "$X") ; [6809909406030619949][1]
              (ConceptNode "compound-A") ; [5571266438078870282][1]
            ) ; [9339663655215837098][1]
          ) ; [12459373116713205302][1]
        ) ; [18413871478668594639][1]
      ) ; [15180311865663055677][1]
      (ExecutionOutputLink
        (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
        (ListLink
          (ImplicationLink
            (PredicateNode "take-treatment-1") ; [693477798751519767][1]
            (VariableNode "$B-a083600") ; [1075803195405159200][3]
          ) ; [16288987656109171854][3]
          (ImplicationLink
            (PredicateNode "take-treatment-1") ; [693477798751519767][1]
            (VariableNode "$B-79625a3") ; [7102392297705646207][3]
          ) ; [13092204721554883053][3]
          (ExecutionOutputLink
            (GroundedSchemaNode "scm: equivalence-to-implication-formula") ; [657362063185547793][1]
            (ListLink
              (ImplicationLink
                (VariableNode "$B-79625a3") ; [7102392297705646207][3]
                (VariableNode "$B-a083600") ; [1075803195405159200][3]
              ) ; [15645609273935500790][3]
              (EquivalenceLink
                (VariableNode "$B-a083600") ; [1075803195405159200][3]
                (VariableNode "$B-79625a3") ; [7102392297705646207][3]
              ) ; [16543044446076083947][3]
            ) ; [16741632124721292660][3]
          ) ; [12795200503087640331][3]
        ) ; [17555647363681032921][3]
      ) ; [17020773718331346226][3]
      (ExecutionOutputLink
        (GroundedSchemaNode "scm: deduction-formula") ; [201750363263560339][1]
        (ListLink
          (ImplicationLink
            (VariableNode "$B-a083600") ; [1075803195405159200][3]
            (LambdaLink
              (TypedVariableLink
                (VariableNode "$X") ; [6809909406030619949][1]
                (TypeNode "ConceptNode") ; [3788388250666256890][1]
              ) ; [17718894144153964593][1]
              (EvaluationLink
                (PredicateNode "take") ; [3396821867884253757][1]
                (ListLink
                  (VariableNode "$X") ; [6809909406030619949][1]
                  (ConceptNode "compound-A") ; [5571266438078870282][1]
                ) ; [9339663655215837098][1]
              ) ; [12459373116713205302][1]
            ) ; [18413871478668594639][1]
          ) ; [9350305881523605350][3]
          (ExecutionOutputLink
            (GroundedSchemaNode "scm: implication-implicant-distribution-formula") ; [4181613870801572780][1]
            (ListLink
              (ImplicationLink
                (VariableNode "$B-a083600") ; [1075803195405159200][3]
                (AndLink
                  (VariableNode "$B-a083600") ; [1075803195405159200][3]
                  (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
                ) ; [17509916884263908246][3]
              ) ; [17669723323973694765][3]
              (ExecutionOutputLink
                (GroundedSchemaNode "scm: implication-construction-formula") ; [6458395065017447120][1]
                (ListLink
                  (ImplicationLink
                    (VariableNode "$B-a083600") ; [1075803195405159200][3]
                    (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
                  ) ; [18429495633157914854][3]
                  (VariableNode "$B-a083600") ; [1075803195405159200][3]
                  (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
                ) ; [11370572920824065288][3]
              ) ; [14390789622547574590][3]
            ) ; [16821536694469753822][3]
          ) ; [9271578245052840560][3]
          (ImplicationLink
            (AndLink
              (VariableNode "$B-a083600") ; [1075803195405159200][3]
              (VariableNode "$Q-1d91711a") ; [9046317156593352527][3]
            ) ; [17509916884263908246][3]
            (LambdaLink
              (TypedVariableLink
                (VariableNode "$X") ; [6809909406030619949][1]
                (TypeNode "ConceptNode") ; [3788388250666256890][1]
              ) ; [17718894144153964593][1]
              (EvaluationLink
                (PredicateNode "take") ; [3396821867884253757][1]
                (ListLink
                  (VariableNode "$X") ; [6809909406030619949][1]
                  (ConceptNode "compound-A") ; [5571266438078870282][1]
                ) ; [9339663655215837098][1]
              ) ; [12459373116713205302][1]
            ) ; [18413871478668594639][1]
          ) ; [16720479476285327004][3]
        ) ; [13021334751802866725][3]
      ) ; [12486461106453180030][3]
    ) ; [9452905340003334304][3]
  ) ; [18141403731508423417][3]
) ; [15511925606899762288][3]
```

I can see they are not alpha equivalent because they have different inference trees

```
 [1075803195405159200][3] [8930023734698229664][3]
 --------implication-construction-formula---------
             [18313202211262791991][3]
             -implication-implicant-d-
             [17553429902078571902][3] [13870626034087997381][3]
             -----------------deduction-formula-----------------
[16346461327915047267][3] [10338135361865330206][3] [14585749684933467565][3]
-equivalence-to-implicat- -----------------deduction-formula-----------------
[16288987656109171854][3]              [9350305881523605350][3]
-----------------------deduction-formula-----------------------
                   [15180311865663055677][1]
```

and

```
                                        [1075803195405159200][3] [9046317156593352527][3]
                                        --------implication-construction-formula---------
                          [16543044446076083947][3] [18429495633157914854][3]
                          -equivalence-to-implicat- -implication-implicant-d-
[13092204721554883053][3] [15645609273935500790][3] [17669723323973694765][3] [16720479476285327004][3]
-----------------deduction-formula----------------- -----------------deduction-formula-----------------
             [16288987656109171854][3]                           [9350305881523605350][3]
             -----------------------------deduction-formula------------------------------
                                      [15180311865663055677][1]
```

**Comment by ngeiswei** [comment]
@linas I'm looking into it but if you happen to have some idea, it's obviously welcome!

**Comment by ngeiswei** [comment]
It seems the problem comes from `ScopeLink::term_hash` which is too destructive, as it returns the same value for the two conjunctions of clauses, and the same value for the two rewrite terms.

**Comment by ngeiswei** [comment]
BTW, @linas , I've defined a `CHandle`, for contextual handle here https://github.com/opencog/atomspace/blob/master/opencog/atomutils/Unify.h#L94 that I think might be useful in places like `ScopeLink::term_hash`, just so you know...

**Comment by ngeiswei** [comment]
I fixed it by replacing
https://github.com/opencog/atomspace/blob/master/opencog/atoms/core/ScopeLink.cc#L330
by
```c++
		hsh += mixer * (hsh<<5) + term_hash(ho, bound_vars, quotation);
```
that is the bit shifting occurs on hsh instead of 1, as you have done elsewhere in the code when aggregating hashes of outgoings.

I do understand the role of the mixer when the link is unordered, I'm a bit concerned however that it is so destructive, if the scope had only an unordered body, the change above wouldn't fix it. What about instead of adding (I mean when mixer is false), we calculate the term_hash of each outgoing, order them, and then apply the same aggregation (now fixed) as if it were ordered? I think it would work (though I'm not very experienced with hash algorithms). What do you say?

**Comment by ngeiswei** [comment]
Manufacturing a clash for unordered links after the proposed fix is proving to be very difficult. I don't know how to find 2 different unordered pairs of atoms that would sum up to the same value. So, since I can't produce a unit test for it, I'm tempted to not touch that function any further. Yet I am a bit worried that this might bite up in the future...

**Comment by linas** [comment]
Nil you asked:

> What about instead of adding (I mean when mixer is false), we calculate the term_hash 
> of each outgoing, order them, and then apply the same aggregation (now fixed) as if it 
> were ordered?

Yes, if done correctly. The trick is that the names of the bound variables need to be excluded from the hashes -- only then can you sort.  As it happens, the `term_hash` function already does this, so, yes, it is enough to replace the no-mixer line (circa lines 330-340 in the current code) by a sorted copy of the hashes.

Yes, that is a good idea, something I should/could have thought of when originally creating the code.  It would remove almost all remaining hash collisions in AndLinks, OrLinks.

**Comment by linas** [comment]
I added a comment lines 330-340 reminding that the above should be done.

**Comment by ngeiswei** [comment]
When rethinking about that, one way to manufacture such a clash would be to build trees composed mostly of unordered links, like `(AndLink (OrLink (SetLink ...)))`. By randomly generating these I'm sure we'd find 2 different ones with same hash not before long, and if we don't, then we probably don't need to worry about it.

**Comment by ngeiswei** [comment]
Oh, BTW, this fix has SUBSTANTIALLY sped up (at least x3) my current BC experiment, by apparently speeding up the pattern matcher (I'm not sure since I didn't log timestamps, just based on watching the stuff running real-time).

**Comment by linas** [comment]
yes, the fix #1177 seems to have fixed a typo/thinko in the code.
